src/circuit/circuit_frame.rs

```diff
@@ -1824,17 +1888,17 @@ fn reduce_cons<F: LurkField, CS: ConstraintSystem<F>>(
         // Otherwise, more than one argument was provided (FIXME: error-checking to ensure exactly two were provided?):
         // [binop_cont_tag, g.op2_eval_tag, env.tag(), env.hash(), more.tag(), more.hash(), cont.tag(), cont.hash()]
 
-        let the_op = pick(
+        let the_op = pick_const(
             &mut cs.namespace(|| "eval op"),
             &end_is_nil,
-            &g.unop_cont_tag,
-            &g.binop_cont_tag,
+            ContTag::Unop.as_field(),
+            ContTag::Binop.as_field(),
         )?;
-        let op1_or_op2 = pick(
+        let op1_or_op2 = pick_const(
             &mut cs.namespace(|| "op1 or op2"),
             &end_is_nil,
-            &g.op1_eval_tag,
-            &g.op2_eval_tag,
+            Op1::Eval.as_field(),
+            Op2::Eval.as_field(),
```

Comment:
Most of those (it should be all, but I might have missed something) still have to be allocated. With a lot more work, it might be possible to change the underlying usages to accept constants, but as far as I know there's not low-hanging fruit. Doing this would only be worth it if it can eliminate *all* allocated usages, since we only get the savings if we can remove from the global allocations. I think this will likely get addressed as part of large upcoming refactors to how all this works (rather than by incremental changes from the current structure).